File: web-frontend/src/types/marketRule.ts
export type SettlementMethod = 'spot' | 'contract' | 'hybrid'
export type DeviationFormulaType = 'stepped' | 'proportional' | 'bandwidth'

export interface SteppedStep {
  lower: number
  upper: number
  rate: number
}

export interface SteppedParams {
  exemption_ratio: number
  steps: SteppedStep[]
}

export interface ProportionalParams {
  exemption_ratio: number
  base_rate: number
}

export interface BandwidthParams {
  bandwidth_percent: number
  penalty_rate: number
}

export type DeviationFormulaParams = SteppedParams | ProportionalParams | BandwidthParams

export interface MarketRuleRead {
  id: string
  province: string
  price_cap_upper: number
  price_cap_lower: number
  settlement_method: SettlementMethod
  deviation_formula_type: DeviationFormulaType
  deviation_formula_params: DeviationFormulaParams
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface MarketRuleCreate {
  price_cap_upper: number
  price_cap_lower: number
  settlement_method: SettlementMethod
  deviation_formula_type: DeviationFormulaType
  deviation_formula_params: Record<string, unknown>
}

export interface DeviationTemplate {
  province: string
  deviation_formula_type: string
  default_params: Record<string, unknown>
}

export const settlementMethodLabels: Record<SettlementMethod, string> = {
  spot: '现货结算',
  contract: '合同结算',
  hybrid: '混合结算',
}

export const deviationFormulaTypeLabels: Record<DeviationFormulaType, string> = {
  stepped: '阶梯式',
  proportional: '比例式',
  bandwidth: '带宽式',
}

File: api-server/app/schemas/market_rule.py
from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.station import Province

SettlementMethod = Literal["spot", "contract", "hybrid"]
DeviationFormulaType = Literal["stepped", "proportional", "bandwidth"]


# --- 偏差考核公式参数 schemas ---


class SteppedStep(BaseModel):
    lower: float = Field(..., ge=0, lt=1)
    upper: float = Field(..., gt=0, le=1)
    rate: float = Field(..., gt=0)

    @model_validator(mode="after")
    def validate_lower_less_than_upper(self) -> "SteppedStep":
        if self.lower >= self.upper:
            raise ValueError("阶梯下限必须小于上限")
        return self


class SteppedParams(BaseModel):
    exemption_ratio: float = Field(..., ge=0, lt=1)
    steps: list[SteppedStep] = Field(..., min_length=1)


class ProportionalParams(BaseModel):
    exemption_ratio: float = Field(..., ge=0, lt=1)
    base_rate: float = Field(..., gt=0)


class BandwidthParams(BaseModel):
    bandwidth_percent: float = Field(..., ge=0, lt=1)
    penalty_rate: float = Field(..., gt=0)


# --- 市场规则 CRUD schemas ---


class MarketRuleCreate(BaseModel):
    price_cap_upper: Decimal = Field(..., gt=0, decimal_places=2)
    price_cap_lower: Decimal = Field(..., ge=0, decimal_places=2)
    settlement_method: SettlementMethod
    deviation_formula_type: DeviationFormulaType
    deviation_formula_params: dict

    @model_validator(mode="after")
    def validate_price_cap_range(self) -> "MarketRuleCreate":
        if self.price_cap_upper <= self.price_cap_lower:
            raise ValueError("最高限价必须大于最低限价")
        return self



class MarketRuleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    province: str
    price_cap_upper: Decimal
    price_cap_lower: Decimal
    settlement_method: SettlementMethod
    deviation_formula_type: DeviationFormulaType
    deviation_formula_params: dict
    is_active: bool
    created_at: datetime
    updated_at: datetime
